### PR TITLE
Persist attributes before save

### PR DESCRIPTION
**Pull request by finalwharf**
What does this PR do?
--------------------
* Allow a user to specify when to persist vault attributes.

Where should the reviewer start?
-------------------------------
* `lib/vault/encrypted_model.rb`

Any background context you want to provide?
------------------------------------------
* Persisting Vault attributes on an `after_save` uses two separate queries:
one for the model `INSERT/UPDATE`, and another to `UPDATE` the ciphertext
for the encrypted attributes. Encrypting the attributes with a `before_save`
avoids the second query. In some cases users might _not_ want to have two
queries when saving a single record. This would be necessary for example,
when one has an auditing table and/or stored procedures that take some action
when a record is changed.

/cc @popovm

### CODE REVIEW

File: spec/unit/encrypted_model_spec.rb [hunk 1]
```diff
@@ -42,4 +42,49 @@
...
+        persist_callback = save_callbacks.first
+
+        expect(persist_callback).to be_a ActiveSupport::Callbacks::Callback
+
+        expect(persist_callback.kind).to eq :after
+      end
+    end
+
+    context "when used" do
+      it "the model has a before_save callback" do
```
Comment (bliof) [inline spec/unit/encrypted_model_spec.rb]: Could you check that `__vault_persist_attributes` is skipped?

Comment (finalwharf) [reply spec/unit/encrypted_model_spec.rb]: The `before_save` and `after_save` callbacks are stored in `_save_callbacks`, which is an `ActiveSupport::Callbacks::CallbackChain` object. The `select` block below filters the callbacks by name, not by `kind` (`:before`,  `:after`, etc.). By ensuring there's only one callback in `_save_callbacks` chain:
```ruby
expect(save_callbacks.length).to eq 1
```
and that it's `kind` is `:before`:
```ruby 
expect(persist_callback.kind).to eq :before
```
and it's name is `:__vault_encrypt_attributes!` (ensured by `filter`), we can be certain that there's no `after_save` callback in the chain with the same name.

Comment (finalwharf) [reply spec/unit/encrypted_model_spec.rb]: Done

**Comment by finalwharf** [comment]
@popovm @bliof The requested changes were addressed. Please review again.

**Pull request merged**